usdt-impl/src/no-linker.rs

```diff
@@ -91,24 +97,25 @@ fn compile_probe(
                     "990:   clr rax",
                     #is_enabled_rec,
                     out("rax") is_enabled,
-                    options(nomem, nostack, preserves_flags)
+                    options(nostack)
                 );
             }
 
             if is_enabled != 0 {
                 #unpacked_args
+                #type_check_fn
                 unsafe {
                     ::std::arch::asm!(
                         "990:   nop",
                         #probe_rec,
                         #in_regs
-                        options(nomem, nostack, preserves_flags)
+                        options(nostack)
```

Comment:
my read of `nomem` and `preserves_flags` is that they're appropriate here; is there a reason you removed them?

Reply:
I was being cautious. The flags might be preserved, I think, but I'm not sure about memory accesses. The raw assembly we've written doesn't read or write any. But let's suppose there's a `char *` in one of the registers, because we're passing a string as a probe argument. DTrace will need to access that memory, but I don't really know if that counts here.

Reply:
> I was being cautious. The flags might be preserved, I think, [...]

Yeah. It's really the nop or the mov so flags are preserved.

> but I'm not sure about memory accesses. The raw assembly we've written doesn't read or write any. But let's suppose there's a `char *` in one of the registers, because we're passing a string as a probe argument. DTrace will need to access that memory, but I don't really know if that counts here.

I don't think that's going to count. Say a debugger attaches to this process when the %rip is on this specific instruction; it then does loads and stores. That's fine.

> The asm! block does not read from or write to any memory accessible outside of the asm! block. This allows the compiler to cache the values of modified global variables in registers across the asm! block since it knows that they are not read or written to by the asm!. The compiler also assumes that this asm! block does not perform any kind of synchronization with other threads, e.g. via fences.

Yeah, we're not going to screw up cached loads for example...

Reply:
Ok, I'll put these flags back.

Reply:
If I'm understanding correctly, then the `nomem` option is also fine in the first ASM block too, right? That's the `clr %rax` bit, which does potentially touch the flags, but does not access any memory at all.